kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,416 @@
+
+/**
+ * The endpoint registry is responsible for associating network endpoints with broker/bootstrap addresses of virtual clusters.
+ * Pictorially the registry looks like this:
+ *
+ * <pre><code>
+ * Registry
+ *    ├─ Endpoint (tls 9092)
+ *    │    ╰───┬──→ cluster1.kafka.com         ──→ Virtual Cluster A bootstrap
+ *    │        ├──→ broker1-cluster1.kafka.com ──→ Virtual Cluster A broker node 1
+ *    │        ╰──→ broker2-cluster1.kafka.com ──→ Virtual Cluster A broker node 2
+ *    ├─ Endpoint (plain 19092)
+ *    │        ╰───→  (null)                   ──→ Virtual Cluster B bootstrap
+ *    ├─ Endpoint (plain 19093)
+ *    │        ╰───→  (null)                   ──→ Virtual Cluster B broker 1
+ *    ╰─ Endpoint (plain 12094)
+ *             ╰───→  (null)                   ──→ Virtual Cluster B broker 2
+ * </code></pre>
+ *
+ *  Key points about the implementation:
+ *  <ul>
+ *    <li>The registry does not take direct responsibility for binding and unbinding network sockets.  Instead, it exposes
+ * a queue of network binding operation events {@link NetworkBindingOperation}.  Callers get the next network
+ * event by calling the {@link #takeNetworkBindingEvent()}.  Once the underlying network operation is complete, the caller
+ * must complete the provided future.</li>
+ *    <li>The registry exposes methods for the registration {@link #registerVirtualCluster(VirtualCluster)} and deregistration
+ * {@link #deregisterVirtualCluster(VirtualCluster)} of virtual clusters.  The registry emits the required network binding
+ * operations to expose the virtual cluster to the network.  These API calls return futures that will complete once
+ * the underlying network operations are completed.</li>
+ *    <li>The registry provides an {@link VirtualClusterBindingResolver}.  The {@link VirtualClusterBindingResolver#resolve(Endpoint, String)} method accepts
+ * connection metadata (port, SNI etc) and resolves this to a @{@link VirtualClusterBinding}.  This allows
+ * Kroxylicious to determine the destination of any incoming connection.</li>
+ * </ul>
+ * The registry is thread safe for all operations.
+ * <ul>
+ *    <li>virtual cluster registration uses java.util.concurrent features to ensure registration is single threaded.</li>
+ *    <li>virtual cluster de-registration uses java.util.concurrent.atomic to ensure de-registration is single threaded.</li>
+ *    <li>updates to the binding mapping (attached to channel) are made only whilst holding an intrinsic lock on the {@link ListeningChannelRecord}.</li>
+ *    <li>updates to the binding mapping are published safely to readers (i.e. threads calling {@link VirtualClusterBindingResolver#resolve(Endpoint, String)}).  This relies the
+ *    fact that the binding map uses concurrency safe data structures ({@link ConcurrentHashMap} and the exclusive use of immutable objects within it.</li>
+ * </ul>
+ */
+public class EndpointRegistry implements AutoCloseable, VirtualClusterBindingResolver {
```

Comment:
As mentioned elsewhere (slack I think). Extracting an interface here seems like a good idea something like 

```java
public interface EndpointRegistry extends AutoCloseable, EndpointResolver {
    CompletionStage<Endpoint> registerVirtualCluster(VirtualCluster virtualCluster);

    CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualCluster);
}
```

I like the idea of keeping the Resolve and the Registry interfaces separate as they perform different roles and in theory are visible to different audiences. 

My personal preference is then to have a `CompositeEndpointRegiistry` which can delegate the registration and resolution to specific instances. My thoughts are something along the lines of `VirtualHostEndpointRegistry` and `DedicatedHostEndpointRegistry`. I was originally thinking of `SniEnpointRegistiry`, `ProxyProtocolEndpointRegistry` and `DecicatedEndpointRegistiry` but the SNI and ProxyProtocol are identical I think its the calling code that would be aware of those destinations.

I'm not quite sure how the composite registry would dispatch register and de-register calls. Does it just forward the to each registry? What would it do if N registries accepted the registration? Should it check if the cluster is TLS enabled? That implies it has very specific knowledge of the properties of the implementations and possibly doesn't work in the proxy protocol case...

Reply:
I'll introduce an interface but keep a single registry implementation for now.

I don't see why `ProxyProtocol` would ever need to be a concern of the of `EndpointRegistry`.  `ProxyProtocol` is just giving us the client IP and that information plays no part in the routing to virtual clusters their broker.   I think that `ProxyProtocol` can remain a concern of the `KafkaProxyFrontendHandler`. Eventually KafkaProxyBackendHandler would need to be prepared to propagate the proxy protocol header to the upstream.

I'm not sure how dispatch would work in a composite registry.  tbh, I'm not sure I think a composite registry would make things simpler.